### PR TITLE
[log-shipper] multiline parser `Custom` type with user provided regex

### LINKED ISSUE #1342: [log-shipper] multiline parser Custom type
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

It is hard to cover all possible multiline variants because their number is only limited by people's imagination.

### Proposed Solution

Provide the Custom multiline parse type with the VRL support.

Example:
```yaml
multiparser:
  type: Custom
  custom:
    starts_when: |
      matched, err = match(.message, r'^\{'); 
      if err != null { false; } else { matched; } 
```
We have a config validation feature so it should not be a problem.

### Additional Information

https://deckhouse.io/en/documentation/v1/modules/460-log-shipper/cr.html#clusterloggingconfig-v1alpha1-spec-multilineparser

### PR DESCRIPTION
**Pull request by alex123012**
## Description

Add new multiline parser type and give users ability to pass their own regexes for multiline log parsing 
## Why do we need it, and what problem does it solve?

Fix #1342 
## What is the expected result?

New multiline parser `Custom` type with user provided regex parsing
## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: log-shipper
type: feature
summary: multiline parser `Custom` type with user provided regex
impact_level: default
```

### CODE REVIEW

File: modules/460-log-shipper/hooks/internal/vector/transform/multiline.go [hunk 1]
```diff
@@ -43,16 +50,61 @@ func CreateMultiLineTransforms(multiLineType v1alpha1.MultiLineParserType) []api
...
+
+	var resultRegexRule vrl.Rule
+	var multilineRegex string
+	if parserRegex.NotRegex != nil {
+		resultRegexRule = vrl.RegexMultilineRule
+		multilineRegex = *parserRegex.NotRegex
+	} else if parserRegex.Regex != nil {
+		resultRegexRule = vrl.RegexMultilineRule
+		multilineRegex = *parserRegex.Regex
+	}
```
Comment (nabokihms) [inline modules/460-log-shipper/hooks/internal/vector/transform/multiline.go]:
```suggestion
	switch {
	case parserRegex.NotRegex != nil:
		multilineRegex = *parserRegex.NotRegex
	case parserRegex.Regex != nil:
		multilineRegex = *parserRegex.Regex 
    default: 
        return fmt.Errorf("regex or notRegex should be provided")
	}
```

**Review by nabokihms — CHANGES_REQUESTED** [review]

File: modules/460-log-shipper/crds/doc-ru-cluster-logging-config.yaml [hunk 2]
```diff
@@ -109,6 +109,26 @@ spec:
...
+                            regex:
+                              description: регулярное выражение, которое считает мэтчем строки, попавшие в него
+                        endsWhen:
+                          description: |
+                            Условие, при котором находится последняя строчка многострокового лога.
+                          properties:
+                            notRegex:
+                              description: регулярное выражение, которое считает мэтчем строки, НЕ попавшие в него
+                            regex:
+                              description: регулярное выражение, которое считает мэтчем строки, попавшие в него
```
[the remark this answers is not in the document]
Comment (nabokihms) [reply modules/460-log-shipper/crds/doc-ru-cluster-logging-config.yaml:131]: And all other dots. Also the sentence should start with a capital letter.

File: modules/460-log-shipper/hooks/internal/vector/transform/multiline.go [hunk 3]
```diff
@@ -43,16 +50,61 @@ func CreateMultiLineTransforms(multiLineType v1alpha1.MultiLineParserType) []api
...
+	return nil
+}
+
+func processMultilineRegex(parserRegex *v1alpha1.ParserRegex) (*string, error) {
+	if parserRegex == nil || (parserRegex.NotRegex == nil && parserRegex.Regex == nil) {
+		return nil, fmt.Errorf("no regex provided")
+	}
+	if parserRegex.NotRegex != nil && parserRegex.Regex != nil {
+		return nil, fmt.Errorf("must be set one of regex or notRegex")
+	}
```
[the remark this answers is not in the document]
Comment (alex123012) [reply modules/460-log-shipper/hooks/internal/vector/transform/multiline.go]: I think, that we should check, that only one of regex or notRegex is provided for parser:
https://github.com/deckhouse/deckhouse/pull/4247/files#diff-f96ecfc87fb685b19794447432b09a05bb4909d50d52192e24eb75d915701910R61

**Review by nabokihms — DISMISSED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @alex123012. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @alex123012. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Pull request merged**